### PR TITLE
Add 'now' feature so one can access the snapshot/chord date

### PR DESCRIPTION
**Pull request by HuwCampbell**
@amosr This is what I have come up with for now (puns!)
We still need to expose DaysDifference in source so it's actually useful.

### CODE REVIEW

File: src/Icicle/Source/ToCore/Base.hs [hunk 1]
```diff
@@ -90,7 +91,8 @@ red n x = mempty { reduces = [(n,x)] }
 post :: Name n -> C.Exp a n -> CoreBinds a n
 post n x = mempty { postcomps = [(n,x)] }
 
-
+pstdt :: Maybe (Name n) -> CoreBinds a n
+pstdt n = mempty { postdate = n }
```
Comment (tranma) [inline src/Icicle/Source/ToCore/Base.hs]: haha, unix naming scheme: 1. remove all vowels, 2. remove all useful letters, 3. repeat until un-pronounceable

File: src/Icicle/Source/ToCore/ToCore.hs [hunk 2]
```diff
@@ -80,8 +80,12 @@ convertQueryTop feats qt
                   Just t' -> return t'
         let inpTy'dated = T.PairT inpTy T.DateTimeT
 
-        (bs,ret) <- evalStateT (convertQuery $ query qt) (ConvertState inp inpTy'dated fs Map.empty)
-        let bs'   = strm inp C.Source <> bs
+        (now,bs,ret) <- evalStateT (do
```
Comment (tranma) [inline src/Icicle/Source/ToCore/ToCore.hs]: blank space

Comment (amosr) [inline src/Icicle/Source/ToCore/ToCore.hs]: lol

Comment (HuwCampbell) [inline src/Icicle/Source/ToCore/ToCore.hs]: https://www.youtube.com/watch?v=e-ORhEE9VVg

Comment (amosr) [inline src/Icicle/Source/ToCore/ToCore.hs]: last night I was talking about taylor swift and laurel said "I've never heard a taylor swift song" and I said "neither have I". no longer can I make that claim, and I hold grudges for a long time

Comment (HuwCampbell) [inline src/Icicle/Source/ToCore/ToCore.hs]: lol

Comment (domdere) [inline src/Icicle/Source/ToCore/ToCore.hs]: its ok @amosr  misery loves company and youtube vids are just so easy to trick people into watching....

File: src/Icicle/Source/ToCore/Base.hs (hunk 1 above)
Comment (amosr) [inline src/Icicle/Source/ToCore/Base.hs]: post straumatic tress disorder

Comment (HuwCampbell) [inline src/Icicle/Source/ToCore/Base.hs]: it was named after `strm`.

Comment (amosr) [inline src/Icicle/Source/ToCore/Base.hs]: I sure do love a slice of humble pie

Comment (tranma) [inline src/Icicle/Source/ToCore/Base.hs]: slhblp

**Comment by amosr** [comment]
this looks good, but I'm not 100% sure about the monoid instance on postdates. what happens if both are "Just"? maybe you'll need to add a new postcomputation binding, setting the second date to the first?

hm can you add now to the arbitrary instance for Source? might be easier to change the Nothing in fets to just some arbitrary variable

**Comment by HuwCampbell** [comment]
Yeah, Monoids on Maybes is a bit terrible. The only real thing we could do it fake a let binding so the next value points to the real one. It's a bit dumb though.
Arbs added.

**Comment by amosr** [comment]
yeah - I think adding lets is good enough

**Comment by HuwCampbell** [comment]
BTW, this PR is on top of the Tombstone one.

**Comment by HuwCampbell** [comment]
Hmm. I don't think it's too simple. The Monoid is used after the convertQuery is called. So it'd be pretty hard to reach back in and add a Let.

**Comment by amosr** [comment]
I was thinking something like
```
mappend (CoreBinds a b c d (Just date1)) (CoreBinds f g h i (Just date2))
 = CoreBinds (a<>f) (b<>g) (c<>h) ((date2, X.XVar date1) : d<>i) (Just date1)	
```
but the other cases same.

ah! but no annotation! damn! maybe you can add an annotation to the CoreBinds (Maybe (a, Date)) field?

**Comment by HuwCampbell** [comment]
Ok, I don't see how it could possibly get called, but I've put it in.

**Comment by amosr** [comment]
ah ok, I misunderstood. I thought each use of "now" was adding a Just to the CoreBinds' date, rather than only being set once at the top level.
yeah - in this case we're better off removing the Maybe PostDate from the CoreBinds, and just adding it as another argument to programOfBinds - since the intermediate steps that also produce CoreBinds don't need to worry about the postdate.

this is good though :microscope:

**Comment by tranma** [comment]
I think postdate shouldn't exist too, but as it is this is still good :hatching_chick: :hatched_chick: :baby_chick: :chicken: :poultry_leg:

**Comment by HuwCampbell** [comment]
Ahh, yes, good point @amosr. Functions with arguments, love it :P

**Comment by amosr** [comment]
tran is currently fixing the mess I created of the typechecker by refusing to use function arguments. oh boy :-(

**Comment by HuwCampbell** [comment]
Just going to close this and create a new PR on top of master.

**Comment by tranma** [comment]
"fix" is a strong word...

**Pull request closed without merging**